### ISSUE TITLE
Unexpected `resource X violates plan` during interactive `pulumi up` related to Archive

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#11587]
When performing a `pulumi up` with `PULUMI_EXPERIMENTAL=true`, I saw the following error:

```
    error: resource urn:pulumi:production::aws-account-setup::pulumi-eng:apps:AccessKeyRotator$aws:lambda/function:Function::rotator-ci-pulumi-service-pulumi-deploy-lambda violates plan: properties changed: ~~code[{&{ d12b4d63b3797584e6d74bb09da2ee1219698efd677800678744960201a2a528 map[__index.js:0xc0033c9d10 node_modules/axios:0xc0033c9d60 node_modules/blakejs:0xc0033c9db0 node_modules/follow-redirects:0xc0033c9e00 node_modules/tweetnacl:0xc0033c9e50 node_modules/tweetsodium:0xc0033c9ea0]  }}!={&{ 23333f4988302b4e113a8ad9587d76c65fe21061bbb4f5a090d7249535bd32c9 map[__index.js:0xc003c9b810 node_modules/axios:0xc003c9b860 node_modules/blakejs:0xc003c9b8b0 node_modules/follow-redirects:0xc003c9b900 node_modules/tweetnacl:0xc003c9b950 node_modules/tweetsodium:0xc003c9b9a0]  }}]
```

This was  during an interactive `pulumi up` which had just performed a preview, and where no changes had been made on disk during the `pulumi up`.  

Is there potentially some non-determinism in Archives that can lead to this?

The code involved as part of this deployment is https://github.com/pulumi/home/pull/2440 (private repo).

**Comment by MitchellGerdisch** [comment]
I can produce this issue deterministically with the code below.

`pulumi up` followed by a `yes` or `pulumi up -y` throws the violates plan error.
`pulumi --skip-preview` does not.

I am running pulumi version:
3.69.0 (I also tried 3.68.0 and 3.67.1 with same result)

Pulumi about:
```
CLI          
Version     tried with 3.69.0, 3.68.0 and 3.67.1
Go Version   go1.20.4
Go Compiler  gc

Plugins
NAME    VERSION
aws     5.41.0
python  unknown

Host     
OS       darwin
Version  13.4
Arch     x86_64

TYPE                  URN
pulumi:pulumi:Stack   urn:pulumi:dev2::rds-plan-issue::pulumi:pulumi:Stack::rds-plan-issue-dev2
pulumi:providers:aws  urn:pulumi:dev2::rds-plan-issue::pulumi:providers:aws::default_5_41_0

Found no pending operations associated with dev2

Backend        
Name           pulumi.com

Dependencies:
NAME        VERSION
pip         23.1.2
pulumi-aws  5.41.0
setuptools  67.8.0
wheel       0.40.0
```

---- CODE ----

```
import pulumi
from pulumi_aws import rds

# Creating an RDS MySQL Instance
rds_instance = rds.Instance("my-rds-instance",
    allocated_storage=20,
    engine="mysql",
    engine_version="5.7",
    instance_class="db.t2.micro",
    name="myexampledatabase",
    username="exampleuser",
    password="examplepassword",
    publicly_accessible=True,
    skip_final_snapshot=True,
)

# Export the RDS Instance endpoint
pulumi.export("rds_instance_endpoint", rds_instance.endpoint)
```

**Comment by Frassle** [comment]
> Update plans are used under the hood by default.

Only if PULUMI_EXPERIMENTAL is true.

**Comment by dixler** [comment]
> > Update plans are used under the hood by default.
> 
> Only if PULUMI_EXPERIMENTAL is true.

This is only true for `pulumi up <template>`
https://github.com/pulumi/pulumi/blob/f1e9239ba6982c781c568858a17543b940dc9359/pkg/cmd/pulumi/up.go#L194-L359

The default `pulumi up` hardcodes it to true.
https://github.com/pulumi/pulumi/blob/f1e9239ba6982c781c568858a17543b940dc9359/pkg/cmd/pulumi/up.go#L86-L153

**Comment by Frassle** [comment]
> This is only true for pulumi up <template\> (via GeneratePlan)

That's for generation not for constraining. It's subtle but constraints are only auto-enabled if experimental is set. See https://github.com/pulumi/pulumi/blob/master/pkg/backend/apply.go#L125-L129.

**Comment by dixler** [comment]
If you look closely, the branch is checking the value of the `kind` parameter which is set to `apitype.UpdateUpdate` in the backends for `pulumi up` and directly plumbed to `PreviewThenPrompt`.

## Context:
https://github.com/pulumi/pulumi/blob/f1e9239ba6982c781c568858a17543b940dc9359/pkg/backend/apply.go#L82-L164
https://github.com/pulumi/pulumi/blob/6f14761b1bc3782c8addb0263095a4228ab22b68/pkg/backend/httpstate/backend.go#L976-L980
https://github.com/pulumi/pulumi/blob/6f14761b1bc3782c8addb0263095a4228ab22b68/pkg/backend/filestate/backend.go#L852-L862
https://github.com/pulumi/pulumi/blob/6f14761b1bc3782c8addb0263095a4228ab22b68/pkg/backend/apply.go#L233-L251

**Comment by Frassle** [comment]
Oh 🤦‍♂️ wrong experimental check: https://github.com/pulumi/pulumi/blob/f1e9239ba6982c781c568858a17543b940dc9359/pkg/backend/apply.go#L217-L219

There was _no way_ we were running constraint checks for plans _by default_ we'd be getting user reports about this on a daily basis.

**Comment by dixler** [comment]
> Oh 🤦‍♂️ wrong experimental check: https://github.com/pulumi/pulumi/blob/f1e9239ba6982c781c568858a17543b940dc9359/pkg/backend/apply.go#L217-L219
> 
> 
> 
> There was _no way_ we were running constraint checks for plans _by default_ we'd be getting user reports about this on a daily basis.

In that case only users that use `--yes` will set AutoApprove to true and not setting `--skip-preview` will be running previews and be using plans?

**Comment by Frassle** [comment]
Previews never use plans, they always generate plans.
Currently the only time plans apply constraints is if PULUMI_EXPERIMENTAL is set, in which case all updates (except those that have --skip-preview) will use the plan, or if --plan is used to explicitly set a plan.

**Comment by Frassle** [comment]
> Are we concerned that non-experimental users may have errors due to bugs in the GeneratePlan code?

No. I expect plan generation to be safe. It's just constraint checking that we know has a lot of issues with non-determinism still.
